module/os/linux/spl/spl-zone.c

```diff
@@ -0,0 +1,325 @@
+		error = ENOENT;
+		goto done;
+	}
+
+	list_del(&zd->zd_list);
+	kmem_free(zd, sizeof (*zd) + zd->zd_dsnamelen + 1);
+
+	/* Prune the namespace entry if it has no more delegations. */
+	if (list_empty(&zds->zds_datasets)) {
+		list_del(&zds->zds_list);
+		kmem_free(zds, sizeof (*zds));
+	}
+	error = 0;
+
+done:
+	mutex_exit(&zone_datasets_lock);
+	return (error);
+#else
+	return (ENXIO);
+#endif
+}
+EXPORT_SYMBOL(zone_dataset_detach);
+
+int
+zone_dataset_visible(const char *dataset, int *write)
+{
+	zone_datasets_t *zds;
+	zone_dataset_t *zd;
+	int visible;
+	size_t dsnamelen, zd_len;
+
+	/* Default to read-only, in case visible is returned. */
+	if (write != NULL)
+		*write = 0;
+	if (zone_dataset_name_check(dataset, &dsnamelen) != 0)
+		return (0);
+	if (INGLOBALZONE(curproc)) {
+		if (write != NULL)
+			*write = 1;
+		return (1);
+	}
+
+	mutex_enter(&zone_datasets_lock);
+	visible = 0;
+	zds = zone_datasets_lookup(crgetzoneid(curproc->cred));
+	if (zds == NULL)
+		goto done;
+
+	list_for_each_entry(zd, &zds->zds_datasets, zd_list) {
+		zd_len = strlen(zd->zd_dsname);
+		/*
+		 * The dataset is visible only if it matches one of the
+		 * namespace's entries.  If it does, it's writable if it's
+		 * an exact match for or deeper than a namespace entry,
+		 * otherwise it's read-only.
+		 */
+		if (zd_len > dsnamelen) {
+			visible = bcmp(zd->zd_dsname, dataset,
+			    dsnamelen) == 0 &&
+			    zd->zd_dsname[dsnamelen] == '/';
+		} else {
+			visible = bcmp(zd->zd_dsname, dataset, zd_len) == 0;
+			if (visible && zd_len > dsnamelen)
+				visible = dataset[zd_len] == '/';
+			if (visible) {
+				if (write != NULL)
+					*write = 1;
+			}
+		}
+	}
+
+done:
+	mutex_exit(&zone_datasets_lock);
+	return (visible);
+}
+EXPORT_SYMBOL(zone_dataset_visible);
+
+#if defined(CONFIG_USER_NS)
+static unsigned int
+user_ns_zoneid(struct user_namespace *user_ns)
+{
+	unsigned int r;
+
+#ifdef HAVE_USER_NS_COMMON_INUM
+	r = user_ns->ns.inum;
+#else
+	r = user_ns->proc_inum;
+#endif
+
+	return (r);
+}
+#endif
+
+unsigned int
+global_zoneid(void)
+{
+	unsigned int z = 0;
+
+#if defined(CONFIG_USER_NS)
+	z = user_ns_zoneid(&init_user_ns);
+#endif
+
+	return (z);
+}
+EXPORT_SYMBOL(global_zoneid);
+
+unsigned int
+crgetzoneid(const cred_t *cr)
+{
+	unsigned int r = 0;
+
+#if defined(CONFIG_USER_NS)
+	r = user_ns_zoneid(cr->user_ns);
+#endif
+
+	return (r);
+}
+EXPORT_SYMBOL(crgetzoneid);
+
+boolean_t
+inglobalzone(proc_t *proc)
+{
+
+#if defined(CONFIG_USER_NS)
+	return (proc->cred->user_ns == &init_user_ns);
+#else
+	return (B_TRUE);
+#endif
+}
+EXPORT_SYMBOL(inglobalzone);
+
+int
+spl_zone_init(void)
+{
+	mutex_init(&zone_datasets_lock, NULL, MUTEX_DEFAULT, NULL);
+	INIT_LIST_HEAD(&zone_datasets);
+	return (0);
+}
+
+void
+spl_zone_fini(void)
+{
+	zone_datasets_t *zds;
+	zone_dataset_t *zd;
+
+	/*
+	 * It would be better to assert an empty zone_datasets, but since
+	 * there's no automatic mechanism for cleaning them up if the user
```

Comment:
Does this mean the dataset is still zoned to the nsid after the namespace is destroyed, until the ZFS module is unloaded? If the nsid is then reused by a new namespace, does it erroneously get access to those datasets?

Reply:
@freqlabs: I think you're describing a use-after-free condition, which could be catastrophic for shared hosting environments. Without a cleanup callback from the namespace' destruction, best we could hope for is some sort of `are_all_delegated_ns_present?` check as part of background maintenance operations, a zed mechanism, or something of that nature to try and win the relevant race. Alternatively, if down the line any of our medical records are in a dataset delegated to a container which is shutdown after completing its op cycle, they could be accessible to the next container owner lucky enough to get the same nsid.

Reply:
Or along similar lines, if we took a reference on the namespace with `get_user_ns()` when attaching it we could prevent it from being recycled until we detached it and called `put_user_ns()` (or during module cleanup).  Seems like we'd just need to somehow lookup the namespace with the `nsid`.

Reply:
So far I have not found a way to go from an `nsid` to the `struct user_namespace *ns` required to take a reference